connstr/connstr.go

```diff
@@ -0,0 +1,698 @@
+			MaxSessions:      DefaultPoolMaxSessions,
+			SessionIncrement: DefaultPoolIncrement,
+			MaxLifeTime:      DefaultMaxLifeTime,
+			WaitTimeout:      DefaultWaitTimeout,
+			SessionTimeout:   DefaultSessionTimeout,
+		},
+	}
+
+	var paramsString string
+	dataSourceName = strings.TrimSpace(dataSourceName)
+	if i := strings.IndexByte(dataSourceName, '\n'); i >= 0 {
+		dataSourceName, paramsString = strings.TrimSpace(dataSourceName[:i]), strings.TrimSpace(dataSourceName[i+1:])
+	}
+	var q url.Values
+
+	if !strings.HasPrefix(dataSourceName, "oracle://") {
+		// Not URL
+		q = make(url.Values, 32)
+		if paramsString == "" &&
+			(strings.HasPrefix(dataSourceName, "connectString=") ||
+				strings.Contains(dataSourceName, " connectString=")) {
+			// logfmt-formatted the whole string
+			paramsString, dataSourceName = dataSourceName, ""
+		} else {
+			// Old, or Easy Connect, or anything
+			P.Username, P.Password.secret, dataSourceName = parseUserPassw(dataSourceName)
+
+			uSid := strings.ToUpper(dataSourceName)
+			//fmt.Printf("dataSourceName=%q SID=%q\n", dataSourceName, uSid)
+			if strings.Contains(uSid, " AS ") {
+				if P.IsSysDBA = strings.HasSuffix(uSid, " AS SYSDBA"); P.IsSysDBA {
+					dataSourceName = dataSourceName[:len(dataSourceName)-10]
+				} else if P.IsSysOper = strings.HasSuffix(uSid, " AS SYSOPER"); P.IsSysOper {
+					dataSourceName = dataSourceName[:len(dataSourceName)-11]
+				} else if P.IsSysASM = strings.HasSuffix(uSid, " AS SYSASM"); P.IsSysASM {
+					dataSourceName = dataSourceName[:len(dataSourceName)-10]
+				}
+			}
+			P.ConnectString = dataSourceName
+		}
+	} else {
+		// URL
+		u, err := url.Parse(dataSourceName)
+		if err != nil {
+			return P, errors.Errorf("%s: %w", dataSourceName, err)
+		}
+		if usr := u.User; usr != nil {
+			P.Username = usr.Username()
+			P.Password.secret, _ = usr.Password()
+		}
+		P.ConnectString = u.Hostname()
+		// IPv6 literal address brackets are removed by u.Hostname,
+		// so we have to put them back
+		if strings.HasPrefix(u.Host, "[") && (len(P.ConnectString) <= 1 || !strings.Contains(P.ConnectString[1:], "]")) {
+			P.ConnectString = "[" + P.ConnectString + "]"
+		}
+		if u.Port() != "" {
+			P.ConnectString += ":" + u.Port()
+		}
+		if u.Path != "" && u.Path != "/" {
+			P.ConnectString += u.Path
+		}
+		q = u.Query()
+	}
+
+	if paramsString != "" {
+		// Parse the logfmt-formatted parameters string
+		d := logfmt.NewDecoder(strings.NewReader(paramsString))
+		for d.ScanRecord() {
+			for d.ScanKeyval() {
+				switch key, value := string(d.Key()), string(d.Value()); key {
+				case "connectString":
+					var user, passw string
+					if user, passw, P.ConnectString = parseUserPassw(value); P.Username == "" && P.Password.IsZero() {
+						P.Username, P.Password.secret = user, passw
+					}
+				case "user":
+					P.Username = value
+				case "password":
+					P.Password.secret = value
+				case "onInit", "shardingKey", "superShardingKey":
```

Comment:
Sharding keys can be not just strings. They can be numbers, dates, and RAW (binary) values. Does this handle only strings? Does it handle multiple values?

Reply:
Yes, only strings - before this, you could specify sharding keys only directly.
How should one specify the data type? 
JSON would handle strings and numbers, but nothing else.